modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/ImageMosaicConfigHandler.java

```diff
@@ -894,40 +894,32 @@ public void updateConfiguration(GridCoverage2DReader coverageReader,
-                // has been already marked as heterogeneous
-
-                numberOfLevels = coverageReader.getNumOverviews(inputCoverageName) + 1;
-                boolean needUpdate = false;
-
-                //
-                // Heterogeneousity check
-                //
-                if (numberOfLevels != mosaicConfiguration.getLevelsNum()) {
-                    catalogConfigurationBean.setHeterogeneous(true);
-                    if (numberOfLevels > mosaicConfiguration.getLevelsNum()) {
-                        resolutionLevels = coverageReader.getResolutionLevels(inputCoverageName);
-                        mosaicConfiguration.setLevels(resolutionLevels);
-                        mosaicConfiguration.setLevelsNum(numberOfLevels);
-                        needUpdate = true;
-                    }
-                } else {
-                    final double[][] mosaicLevels = mosaicConfiguration.getLevels();
-                    resolutionLevels = coverageReader.getResolutionLevels(inputCoverageName);
-                    final boolean homogeneousLevels = Utils.homogeneousCheck(numberOfLevels,
-                            resolutionLevels, mosaicLevels);
-                    if (!homogeneousLevels) {
-                        catalogConfigurationBean.setHeterogeneous(true);
-                        needUpdate = true;
-                    }
+
+            // make sure we pick the same resolution irrespective of order of harvest
+            numberOfLevels = coverageReader.getNumOverviews(inputCoverageName) + 1;
+            resolutionLevels = coverageReader.getResolutionLevels(inputCoverageName);
+            
+            int originalNumberOfLevels = mosaicConfiguration.getLevelsNum();
+            boolean needUpdate = false;
+            if (Utils.homogeneousCheck(Math.min(numberOfLevels, originalNumberOfLevels), 
+                    resolutionLevels, mosaicConfiguration.getLevels())) {
+                if (numberOfLevels > originalNumberOfLevels) {
+                    needUpdate = true; // pick the one with highest number of levels
                 }
-                // configuration need to be updated
-                if (needUpdate) {
-                    getConfigurations().put(mosaicConfiguration.getName(), mosaicConfiguration);
+            } else {
+                if (isHigherResolution(resolutionLevels, mosaicConfiguration.getLevels())) {
+                    needUpdate = true; // pick the one with the highest resolution
                 }
             }
+
+            // configuration need to be updated
+            if (needUpdate) {
+                catalogConfigurationBean.setHeterogeneous(true);                
+                mosaicConfiguration.setLevels(resolutionLevels);
+                mosaicConfiguration.setLevelsNum(numberOfLevels);
+                getConfigurations().put(mosaicConfiguration.getName(), mosaicConfiguration);
+            }
+
```

Comment:
I think that with this approach, heterogeneous won't be set to true in this case:
GranuleA_xResolutions: 0.1,  0.2,  0.4  --> (3 levels)
GranuleB_xResolutions: 0.1,  0.2          --> (2 levels)
MosaicConfiguration is already set with GranuleA properties.

Analyzing GranuleB:
  - Utils.homogeneousCheck return true since it uses Math.min
  - numberLevels = 2 < originalNumberOfLevels = 3
Therefore needUpdate is never set to true so no heterogeneous check is set to true.

If granules would have been analyzed in the opposite order (GranuleB, then GranuleA), 
numberOfLevels = 3 > originalNumberOfLevels = 2 will have triggered needUpdate = true.

Wondering if the failing tests on this pull request are due to this.
I have checkout your branch on my local machine and it's failing here too.